proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,262 @@
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+* *3 - downstream errors does not record account for errors such as TLS negotiation errors or failures to determine virtual cluster or broker.  
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+
+## Use Cases
+
+In general, metrics exist to help a user understand the health and performance of the proxy. They help inform
+capacity planning, drive system alerts reporting abnormal conditions and help with problem-solving (things like
+diagnosing failures or system slow-downs).
+
+Introducing a proxy to a Kafka system will increase latency of the system. Users will need a mechanism to understand how
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+* filters can perform asynchronous work.  When asynchronous work occurs, the proxy stops reading the downstream or upstream.
+
+## Proposal
+
+## Summary
+
+The following metrics will be added.  They will each be described in more detail later in this section.
+
+| Metric                                           | Type         | Labels                                                                                | Description                                                                             |
+|--------------------------------------------------|--------------|---------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
+| kroxylicious_downstream_kafka_message_total      | Counter      | virtual_cluster, node_id†, flowing, api_key, api_version, opaque                      | Incremented by one every time a request comes from/response goes to the downstream.     |
+| kroxylicious_upstream_kafka_message_total        | Counter      | virtual_cluster, node_id†, flowing, api_key, api_version, opaque, originating_filter† | Incremented by one every time a request goes to/response comes from the upstream.       |
```

Comment:
I'm sceptical about sometimes including labels. I think it forces people to have to write more complicated post-processing to extract the measurement they actually want. Here when `originating_filter` is present you're measuring a completely different thing. I don't think people are often going to want an aggregate which included the filter-originated request with the client-originated ones. To me it makes much more sense to embed the originator in the metric name.

* `proxy_to_server_request_total`
* `filter_to_server_request_total`
* `server_to_proxy_response_total`
* `server_to_filter_response_total`

Reply:
I like filter_to_server

Reply:
I like the idea if separated metrics for requests sent async by the filter, however filter_to_server not too clear.  A user writing a filter that intercepts requests might expect f2s to tick.

I think we need a need that communicates that these metrics relate to the filter asynchronous request.

filter_async_to_server is naff.

(Thinking...)

Reply:
I decided I could live with _filter_to and _to_filter.  The description will make it clear how they are used.